File: bindings/csharp/src/ProvenFileserver.cs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// File Server protocol types for proven-servers.

namespace Proven;

/// <summary>FileOperation matching the Idris2 ABI tags (0-9).</summary>
public enum FileOperation : byte
{
    Read = 0,
    Write = 1,
    Create = 2,
    Delete = 3,
    Rename = 4,
    List = 5,
    Stat = 6,
    Lock = 7,
    Unlock = 8,
    Watch = 9
}

/// <summary>FileType matching the Idris2 ABI tags (0-6).</summary>
public enum FileType : byte
{
    Regular = 0,
    Directory = 1,
    Symlink = 2,
    BlockDevice = 3,
    CharDevice = 4,
    Fifo = 5,
    Socket = 6
}

/// <summary>FilePermission matching the Idris2 ABI tags (0-8).</summary>
public enum FilePermission : byte
{
    OwnerRead = 0,
    OwnerWrite = 1,
    OwnerExecute = 2,
    GroupRead = 3,
    GroupWrite = 4,
    GroupExecute = 5,
    OtherRead = 6,
    OtherWrite = 7,
    OtherExecute = 8
}

/// <summary>LockType matching the Idris2 ABI tags (0-3).</summary>
public enum LockType : byte
{
    Shared = 0,
    Exclusive = 1,
    Advisory = 2,
    Mandatory = 3
}

/// <summary>FileErrorCode matching the Idris2 ABI tags (0-9).</summary>
public enum FileErrorCode : byte
{
    NotFound = 0,
    PermissionDenied = 1,
    AlreadyExists = 2,
    NotEmpty = 3,
    IsDirectory = 4,
    NotDirectory = 5,
    NoSpace = 6,
    ReadOnly = 7,
    Locked = 8,
    IoError = 9
}

/// <summary>SessionState matching the Idris2 ABI tags (0-4).</summary>
public enum SessionState : byte
{
    Idle = 0,
    Connected = 1,
    Operating = 2,
    FsLocked = 3,
    Disconnecting = 4
}

File: bindings/php/src/ProvenFileserver.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// File Server protocol types for proven-servers.

declare(strict_types=1);

namespace ProvenServers;

/** FileOperation matching the Idris2 ABI tags. */
enum FileOperation: int
{
    case Read = 0;
    case Write = 1;
    case Create = 2;
    case Delete = 3;
    case Rename = 4;
    case List = 5;
    case Stat = 6;
    case Lock = 7;
    case Unlock = 8;
    case Watch = 9;
}

/** FileType matching the Idris2 ABI tags. */
enum FileType: int
{
    case Regular = 0;
    case Directory = 1;
    case Symlink = 2;
    case BlockDevice = 3;
    case CharDevice = 4;
    case Fifo = 5;
    case Socket = 6;
}

/** FilePermission matching the Idris2 ABI tags. */
enum FilePermission: int
{
    case OwnerRead = 0;
    case OwnerWrite = 1;
    case OwnerExecute = 2;
    case GroupRead = 3;
    case GroupWrite = 4;
    case GroupExecute = 5;
    case OtherRead = 6;
    case OtherWrite = 7;
    case OtherExecute = 8;
}

/** LockType matching the Idris2 ABI tags. */
enum LockType: int
{
    case Shared = 0;
    case Exclusive = 1;
    case Advisory = 2;
    case Mandatory = 3;
}

/** FileErrorCode matching the Idris2 ABI tags. */
enum FileErrorCode: int
{
    case NotFound = 0;
    case PermissionDenied = 1;
    case AlreadyExists = 2;
    case NotEmpty = 3;
    case IsDirectory = 4;
    case NotDirectory = 5;
    case NoSpace = 6;
    case ReadOnly = 7;
    case Locked = 8;
    case IoError = 9;
}

/** SessionState matching the Idris2 ABI tags. */
enum SessionState: int
{
    case Idle = 0;
    case Connected = 1;
    case Operating = 2;
    case FsLocked = 3;
    case Disconnecting = 4;
}
